### ISSUE TITLE
[Segment Replication] Update shard allocation to evenly distribute primaries.

### ISSUE THREAD

**Issue by mch2** [opensearch-project/OpenSearch#5240]
With segment replication primary shards will be using more node resources than replicas.  While still net less than docrep, this will lead to an uneven utilization of resources across the cluster.

We need to explore updating allocation with segrep enabled to evenly balance primary shards across a cluster.

- [x] Add primary shard balance setting in the cluster. This is needed to provide mechanism for balancing cluster based on primary shards. https://github.com/opensearch-project/OpenSearch/pull/6017
- [x] Prefer segment replication enabled indices first for allocation and rebalance. This is to avoid skewness of segrep shards on specific nodes. 
- [ ] Perform analysis on reasonable defaults to be used for balanced cluster. 
- [ ] #6210

**Comment by dreamer-89** [comment]
Looking into it

**Comment by dreamer-89** [comment]
Tried a quick exercise to see existing shard allocation behavior. The overall shards (including primary and replica) were distributed evenly across nodes but not primary. This is due to existing [EnableAllocationDecider](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/routing/allocation/decider/EnableAllocationDecider.java#L95) which by default rebalance all shards irrespective whether it is replica or primary. 

Create Index with SEGMENT as replication type. 
```
{
  "settings": {
    "index": {
      "number_of_shards": 5,  
      "number_of_replicas": 1,
      "replication.type": "SEGMENT" 
    }
  }
}
```

```
➜  OpenSearch git:(main) ✗ curl "localhost:9200/_cat/shards?v" | grep "runTask-1" | grep " p " | wc -l
      13
➜  OpenSearch git:(main) ✗ curl "localhost:9200/_cat/shards?v" | grep "runTask-2" | grep " p " | wc -l
      13
➜  OpenSearch git:(main) ✗ curl "localhost:9200/_cat/shards?v" | grep "runTask-0" | grep " p " | wc -l
       9

➜  OpenSearch git:(main) ✗ curl "localhost:9200/_cat/shards?v" | grep "runTask-0"  | wc -l     
      24
➜  OpenSearch git:(main) ✗ curl "localhost:9200/_cat/shards?v" | grep "runTask-1"  | wc -l
      23
➜  OpenSearch git:(main) ✗ curl "localhost:9200/_cat/shards?v" | grep "runTask-2"  | wc -l
      23
```

I retried the exercise after enabling rebalancing on primaries but that did not change the result. 
```
{
  "persistent" : {
    "cluster.routing.rebalance.enable": "primaries"
  }
}
```

**Comment by dreamer-89** [comment]
`cluster.routing.rebalance.enable` gates which shard type can be rebalance but doesn't govern the actual allocation. The existing allocation is based on shard count (irrespective of primary/replica) and is govern by [WeightFunction](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java#L249) inside BalancedShardAllocator

**Comment by dreamer-89** [comment]
There are couple of approaches to solve this: 
### Handle Segment replication enabled shards differently 
Add new parameter in [WeightFunction](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java#L249) `primaryShardBalance` which is used for weight calculation of segment replication enabled shards. 

Pros

-  It does not change existing balancing logic

Cons

- Two different weight calculations in cluster may not result in even balance. 

### Change allocation logic for all type of shards
Change existing WeightFunction for all type of shards. The algorithm assigns higher weight (or introduces another `primaryIndex` balance factor) to ensure even primary shard distribution followed by replicas.

Pros:

- Unifies logic for allocation not just for SegRep but also for RemoteStore and more. 

Cons:

- Changing the core logic may cause unintended regression and behavior change

**Comment by Bukhtawar** [comment]
I think we need to associate default weights per shard based on various criteria f.e primary/replica or remote/local such that these weights are representative of the multi-dimensional work(compute/memory/IO/disk) they do relative to one another. This will ensure we are able to tune these vectors based on the heat dynamically as well(long term), once indices turn read only or requests distribution shifts on different shards based on custom routing logic.

**Comment by dreamer-89** [comment]
> I think we need to associate default weights per shard based on various criteria f.e primary/replica or remote/local such that these weights are representative of the multi-dimensional work(compute/memory/IO/disk) they do relative to one another. This will ensure we are able to tune these vectors based on the heat dynamically as well(long term), once indices turn read only or requests distribution shifts on different shards based on custom routing logic.

Thanks @Bukhtawar for the feedback and the suggestion. This is definitely useful in larger scheme of things with other features and probably need overhaul of existing allocation. I am thinking of starting with updating existing `BalancedShardAllocator` and tune weight function to incorporate segment replication semantics. We can re-iterate and make model more robust in follow up. Please feel free to open an issue with more details and we can discuss more.

**Comment by dreamer-89** [comment]
#### Requirement
Ensure even distribution of same index primaries (and replicas). Replicas also need even distribution because primaries for an index may do variant amount of work (hot shards), so does corresponding replica. 

#### Goals
As part of phase 1, targeting below goals for this task.
1. Shard distribution resulting in overall balance
2. Minimal side effect on rebalancing
3. Ensure no negative impact.
- Transient or permanent unassinged shards (red cluster). 
- Comparatively higher shard movements post allocation & rebalancing. Node join & drop etc

#### Assumption
Balancing primary shards evenly for docrep indices does not impact the cluster performance i.e. no impact on performance. 

#### Approach
As overall goal is to have uniform resource utilization among nodes containing various type of shards, we need to use same scale to measure different shard types. Keeping separate allocation logic for different shard types, working in isolation will NOT work! I think it make sense to update the existing weightFunction factoring in primary shards which solves use case here. This will be the initial work to evolve the existing weighing function. 

#### Future work
The weight function can be updated to cater future needs, where different shards have in-built weight against a dimension (or weighing factor). E.g. primary with more replicas (higher fan out) have higher weight compared to other primaries; so that weight function prioritize even distribution of these primaries.

The ideal weight function should include
1. Primary count/index. Primary shards should be distributed evenly because they do 1:M replication work, where M is no of replicas. 
2. Shard count/index. The weighing function should consider the replica count as well because more number of replicas means more work for primary shard i.e. primary shard with more replicas should have higher weight compared to shard with lesser replicas. 
3. Replica count/index. Shard of specific size will have similar resource consumption. It will be problematic to compare all segrep enabled replicas as same e.g. bigger shard will have high I/O requirements. So, keeping them balanced across nodes is useful. 

I think based on above, the shards can be ordered as below. 
`Primary shard with more replicas > Primary shard > Replica shard`
The weight function then becomes
```
W = w1*f1 + w2*f2 + .... + wn*fn
where wi is current weight on node and fi is the constant defining weight factor. 
```

**Comment by dreamer-89** [comment]
There is no single value for `PRIMARY_BALANCE_FACTOR_SETTING` which distributes shards satisfying all three balance factors for every possible cluster configuration. This is true even today with only INDEX_BALANCE_FACTOR_SETTING and  SHARD_BALANCE_FACTOR_SETTING. A higher value for primary shard balance factor may result in primary balance but not necessarily equal shards per node.

**Comment by dreamer-89** [comment]
### For subtask 2 mentioned [here](https://github.com/opensearch-project/OpenSearch/issues/5240#issue-1448643746)
Previous change in #6017 introduces primary shard balance factor but doesn't differentiate on shard types (docrep vs segrep); which attains overall balanced primary shard distribution but not necessarily for individual shard types. Example, the nodes removal results in 4 primary unassigned shards (docrep 2 and segrep 2), there are chances of one node getting both docrep shard while other getting segrep shards. Changing this logic to accomodate segrep index is not required because:

1. LocalShardsBalancer [sorts](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java#L756) priortises primary first, for allocation which means balanced distribution across nodes. Also, it is not possible to create multiple indices at once, so it always deals with single indices at a time. 
2. LocalShardsBalancer logic is meant for newly created indices only. Failed shard is handled by RoutingNodes#failShard which promotes in-sync replica copy (if any); while existing shard allocator (GatewayAllocator) promotes an upto date version of shard as primary
3. The rebalancing logic is still applied post failover scenario (in step 2) to keep balanced primary shard distribution after #6017

For logging purpose, added an unsuccessful test to mimic scenario where LocalShardsBalancer allocation logic is applied [here](https://github.com/dreamer-89/OpenSearch/blob/balanced_segrep_shards/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationAllocationIT.java#L204)

**Comment by dreamer-89** [comment]
Tracking remaining work in https://github.com/opensearch-project/OpenSearch/issues/6210 for benchmarking. Closing this issue.
